Add orthographic 3d rendering mode

autious opened the pull request:
I needed orthographic rendering for my own things. This solution doesn't modify the existing Camera struct but instead uses a new one called CameraOrtho and a new Begin3dModeOrtho() call. This means that the input systems and picking functions are not compatible, but for that to be possible they would need a lot of internal reimplementation. Some of the semantics might not be compatible either, suggesting that only a partial reimplementation would be necessary, further suggesting that a separate set of functionality is warranted.

This is a (at least partial) solution of issue #439

victorfisac commented:
I think it would be better to implement a ENUM like CAMERA_PERSPECTIVE, CAMERA_ORTHOGRAPHIC. And use just Begin3dMode()/Begin3dMode().

Inside Begin3dMode(), check which value has this enum and initialize 'double top' and use rlFrustum or rlOrtho; because the other code inside of Begin3dModeOrtho() is same as Begin3dMode().

What do you think, @autious @raysan5 ?

autious commented:
I considered that, but it would break all existing code that uses the c-style initializer for the Camera struct and i didn't feel comfortable making a breaking change. This fix will not affect any existing code. I would respect such a change to occur only on the authority of the core team of maintainers.

a3f commented:
> it would break all existing code that uses the c-style initializer for the Camera struct

It would only break ABI (which we intend to do with raylib 2.0, anyway), but code (API) should stay the same. Just add the new `enum` as the last `struct` member and it will automatically be initialized to `0` (which `CAMERA_PERSPECTIVE` should be defined as).

Minor downside would be overzealous warnings (`-Wmissing-field-initializers`).

raysan5 commented:
Agree with @victorfisac and @a3f.

autious commented:
Understood, i'll create a variation of the implementation matching the suggestions.

autious commented:
Modified version uploaded. I haven't tested all the functions that use a Camera as a parameter, but i have modified the Ray cast and world-point projection to mouse coordinate functions, they appear to function correctly in my personal project (with a minor deviation of ~0.1 between each other)

autious commented:
I also created an extra example program that allows the user to switch between perspective and orthogonal rendering, to demonstrate the visual difference. A better variant for student insight might be to have a camera moving around the object and zooming in/out to fully visualize it.

victorfisac commented on src/core.c:
```diff
@@ -919,13 +919,26 @@ void Begin3dMode(Camera camera)
+    if(camera.type == CAMERA_PERSPECTIVE) 
+    {
+        // Setup perspective projection
+        float aspect = (float)screenWidth/(float)screenHeight;
```
Why you initialize aspect inside if/else. It has the same value in both cases.

victorfisac commented on src/core.c:
```diff
@@ -919,13 +919,26 @@ void Begin3dMode(Camera camera)
+        // Setup perspective projection
+        float aspect = (float)screenWidth/(float)screenHeight;
+        double top = 0.01*tan(camera.fovy*0.5*DEG2RAD);
+        double right = top*aspect;
```
Same with this.

victorfisac commented on examples/models/models_orthographic_projection.c:
```diff
@@ -0,0 +1,109 @@
+        // Draw
+        //----------------------------------------------------------------------------------
+
+        if(view_ortho) 
```
This should be executed inside IsKeyPressed condition statement

victorfisac commented on examples/models/models_orthographic_projection.c:
```diff
@@ -0,0 +1,109 @@
+    //--------------------------------------------------------------------------------------
+    int screenWidth = 800;
+    int screenHeight = 450;
+    bool view_ortho = true;
```
You should be using camera->type instead of a global boolean.

autious commented:
Some changes have been made based on suggestions made by @victorfisac

victorfisac commented on examples/models/models_orthographic_projection.c:
```diff
@@ -0,0 +1,109 @@
+
+    InitWindow(screenWidth, screenHeight, "raylib [models] example - geometric shapes");
+
+    const Camera perspective_camera = (Camera){{ 0.0f, 10.0f, 10.0f }, { 0.0f, 0.0f, 0.0f }, { 0.0f, 1.0f, 0.0f }, 45.0f, CAMERA_PERSPECTIVE };
```
You can reduce code amount using just one Camera variable (without const) with perspective values and define values:
```
#define      FOVY_ORTHOGRAPHIC        10.0f
#define      FOVY_PERSPECTIVE         45.0f
```

So, you can then:

```
if(IsKeyPressed(KEY_SPACE)) 
{
     if(camera.type == CAMERA_PERSPECTIVE) 
     {
          camera.fovy = FOVY_ORTHOGRAPHIC;
          camera.type = CAMERA_ORTHOGRAPHIC;
     } 
     else 
     {
          camera.fovy = FOVY_PERSPECTIVE;
          camera.type = CAMERA_PERSPECTIVE;
     }
}
```

raysan5 commented:
Hey @autious! Many thanks for working on this improvement!

@victorfisac thanks for the review and the proposed ideas!

autious commented:
You're welcome. It's always nice to contribute a little.
